File: assignment 4/c/programs/4.c
#include <stdio.h>
#include <malloc.h>

typedef struct node
{
    int item;
    struct node *next;
} Node;

Node *START = NULL;

void create_list(int item)
{
    Node *temp, *ptr;
    temp = (Node *)malloc(sizeof(Node));
    temp->item = item;
    temp->next = NULL;
    if (START == NULL)
    {
        START = temp;
    }
    else
    {
        ptr = START;
        while (ptr->next != NULL)
        {
            ptr = ptr->next;
        }
        ptr->next = temp;
    }
}

void insert_at_specific(int data, int loc)
{
    Node *p;
    Node *temp = (Node *)malloc(sizeof(Node));
    temp->item = data;
    temp->next = NULL;
    int i;
    p = START;
    for (i = 1; i < loc - 1; i++)
    {
        p = p->next;
    }
    temp->next = p->next;
    p->next = temp;
}

void delete_at_end()
{
    Node *p, *q;
    if (START->next == NULL)
    {
        p = START;
        START = NULL;
        free(p);
    }
    else
    {
        p = START;
        while (p->next != NULL)
        {
            q = p;
            p = p->next;
        }
        q->next = NULL;
        free(p);
    }
}

int main()
{
    int n, i, data, a, b, loc, item_to_insert, item_to_delete_at_end;

    printf("Enter the number of items: ");
    scanf("%d", &n);

    for (i = 0; i < n; i++)
    {
        printf("Enter the data item: ");
        scanf("%d", &data);
        create_list(data);
    }

    printf("Enter the location where you want to insert an item: ");
    scanf("%d", &loc);

    printf("Enter the item you want to insert: ");
    scanf("%d", &item_to_insert);

    insert_at_specific(item_to_insert, loc);

    printf("Do you want to delete an item at the end of the list? Enter 1 for yes and 0 for no: ");
    scanf("%d", &item_to_delete_at_end);

    if (item_to_delete_at_end == 1)
        delete_at_end();

    Node *p = START;

    printf("START");
    while (p != NULL)
    {
        printf("-->%d", p->item);
        p = p->next;
    }
    printf("\n");

    return 0;
}

File: assignment 4/python/programs/4.py
class Node:
    def __init__(self, item):
        self.item = item
        self.next = None


START = None


def create_list(item):
    global START
    temp = Node(item)
    temp.next = None
    if START is None:
        START = temp
    else:
        ptr = START
        while ptr.next is not None:
            ptr = ptr.next
        ptr.next = temp


def insert_at_specific(data, loc):
    global START
    temp = Node(data)
    temp.next = None
    p = START
    for i in range(1, loc - 1):
        p = p.next
    temp.next = p.next
    p.next = temp


def delete_at_end():
    global START
    p = START
    q = None
    if p.next is None:
        START = None
    else:
        while p.next is not None:
            q = p
            p = p.next
        q.next = None


n = int(input("Enter the number of items: "))

for _ in range(n):
    data = int(input("Enter the data item: "))
    create_list(data)

loc = int(input("Enter the location where you want to insert an item: "))
item_to_insert = int(input("Enter the item you want to insert: "))
insert_at_specific(item_to_insert, loc)

item_to_delete_at_end = int(
    input(
        "Do you want to delete an item at the end of the list? Enter 1 for yes and 0 for no: "
    )
)
if item_to_delete_at_end == 1:
    delete_at_end()

p = START
print("START", end="")
while p is not None:
    print(f"--> {p.item}", end="")
    p = p.next
print()
